### DISCUSSION TITLE
pagination of external db data

### DISCUSSION THREAD

**Discussion by gadshaked** [Budibase/budibase#2596] [Help] [answered]
Hi,
I am new to Budibase.
Is there some way to do pagination on external db (postgresql)?
I the GUI states that "Pagination is only available for data stored in internal tables." , so I wonder if there is an alternative.
Thank you
Gad

**Comment by mjashanks** [comment, marked as answer]
Hi @gadshaked .

The postgresql connector should handle pagination automatically when using the in-built CRUD capabilities.

In the video below, I setup a data provider that is connected to a Postgres table. I the set the `Limit` property to 2, which allows pagination of the table.

If you wish to paginate a custom SQL query, yo can do so by using setting up a `limit` parameter in your query. Your query would then look like:

`select * from customer limit {{ limitParameter }}`


https://user-images.githubusercontent.com/3524181/133113032-e67910c7-f325-4f04-b04d-a140612acb2c.mp4

**Reply by samg666** [reply]
Hello,

I have the same pagination request for custom sql(MySQL), I tried to set a  limit parameter default to 50, and data provider limit to 2(and Table row to 2) and turned paginate on, then there's still no pagination. The data provider/table still got 2 rows, but the sql actually returned 50 rows. Is there anything I have missed? Or it's not support pagination on custom SQL?

<img width="1872" alt="iShot2021-10-16 08 53 48" src="https://user-images.githubusercontent.com/29649651/137567378-1ba6204b-f014-418a-918c-d2c7867ca092.png"
<img width="755" alt="iShot2021-10-16 08 54 33" src="https://user-images.githubusercontent.com/29649651/137567385-2f05387c-5ac5-4e5c-9bc6-42a1255ed3cd.png">
>

**Comment by gadshaked** [comment]
Hi,
Thank you for your answer.
I tried to not use custom queries, however for some unknown reason I see
only the first row in the database.
If I use a custom query like yours "select * from customer limit {{
limitParameter }}" I get all rows but it only returns the number of rows I
configured in the limit property but pagination is not working.
I am using version 0.9.124.
please advise,
thank you,
Gad

**Reply by mjashanks** [reply]
Hey Gad.

So, just to clarify what you have done so far

- You've created a Postgres data source, and used the "Fetch tables from database" button to pull back all your tables
- In the `Design` section, you've created a `Data Provider` component, that points to one of your tables
- You've added a `Table` component (or a Repeater?) underneath the `Data Provider`

Are the above statements correct?

If so, could you check
1. The `Limit` property on your data provider? What is it?
2. Do you have any filters on your data provider? (i.e. under "Define Filters")

It would be useful if you could provide some screenshots of 
- Your data provider settings
- Your component structure (in the left panel, under screen in question)

Thanks!

**Comment by gadshaked** [comment]
Hi Michael,

   - You've created a Postgres data source, and used the "Fetch tables from
   database" button to pull back all your tables - *Yes I did*
   - In the Design section, you've created a Data Provider component, that
   points to one of your tables -*True*
   - You've added a Table component (or a Repeater?) underneath the Data
   Provider - *True*

Here are the requested screenshots

[image: image.png]

[image: image.png]


Thank you,
Gad

**Comment by gadshaked** [comment]
Also,
I did not setup any filters
Gad

**Comment by gadshaked** [comment]
Hi,
I also turned on log_statement = 'all' on my PG server and it seems that Budibase query is OK, however it just include only the first row of the table


2021-09-26 11:07:40.811 IDT [19520] LOG:  execute <unnamed>: select "tbl_deploy_log"."fname" as "tbl_deploy_log.fname" from "tbl_deploy_log" limit $1
2021-09-26 11:07:40.811 IDT [19520] DETAIL:  parameters: $1 = '50'

Thanks,
Gad

**Reply by mjashanks** [reply]
Hi @gadshaked - I'm not sure what is going on! Would you be able to export your app and email it to me? My address is in my profile.

That will allow me to debug your issue. Thanks!

**Reply by gadshaked** [reply]
Hi Michael,
I emailed you the export file.
Thanks,
Gad

**Reply by mjashanks** [reply]
Thanks Gad. The only thing I can see is that the "Sort Column" on your data provider is not set. 

I think it gets unset when you change the table of the datasource, and the table does not have a column by that name.

However, when I do this, it causes zero rows to be returned, until I change the sort column - so I'm not sure if this is your issue.

The other slightly strange thing is that you have a Form inside your DataProvider. I don't think this is causing an issue to be honest, but it's unusual!

Try removing your data provider, then re-add it (and the table). If it's still an issue, and I'll look further, by recreating your tables in my DB.

**Comment by gadshaked** [comment]
Hi,
The sort column was set to "created_at".
I also tried recreating the data provider and the table under it but I
still get a single row of data.

Thanks,
Gad

**Reply by mjashanks** [reply]
Hi Gad.

I recreated all of your tables in my own DB. I can reproduce the issue. The issue is caused because your tables do not have an identity column. 
Budibase needs an identity column to be able to load single rows into a detail page... e.g. `/deploy_log/:id`.

To get this to work
1. Add an identity column. e.g.

```sql
ALTER TABLE public.tbl_deploy_log
    ADD COLUMN id integer NOT NULL GENERATED ALWAYS AS IDENTITY;
```
2. Click on the "Fetch tables from database" again.

**Reply by mjashanks** [reply]
Also, i logged this as an issue: https://github.com/Budibase/budibase/issues/2820
